### DISCUSSION TITLE
How do I start a span for aggregate context?

### DISCUSSION THREAD

**Discussion by erickskrauch** [open-telemetry/opentelemetry-go#4923] [Q&A]
Hello.

I am using API with a strict Rate Limit. This API has an endpoint that can be called by passing multiple entities there. I have implemented a request aggregation functionality so that all incoming requests are queued, then grouped and sent ([code link](https://github.com/elyby/chrly/blob/f5bc474b4d8bf26c900b84b4c05d6b3b37432210/internal/mojang/batch_uuids_provider.go#L122-L129)).

I would like the outgoing HTTP request to have a span. But I don't understand how I can start 1 span for 10 different contexts. I've seen libraries that allow grouping multiple contexts, but that won't work for otel.

What would be an idiomatic way to implement such a tooling correctly?

**Comment by dmathieu** [comment]
Create the span once. Then, for every context, you can assign the existing span to it with `trace.ContextWithSpan`

https://github.com/open-telemetry/opentelemetry-go/blob/main/trace/context.go#L24

**Reply by erickskrauch** [reply]
That's good to know. But how can I inside an implementation of sending an HTTP request start span from a context, so it doesn't know that there are multiple contexts in here?

It's hard for me to give you an example as it requires a lot of code and it would be easier for you to see the implementation [here](https://github.com/elyby/chrly/blob/f5bc474b4d8bf26c900b84b4c05d6b3b37432210/internal/mojang/batch_uuids_provider.go#L122-L129). The point is that the `fireRequest()` function has a different context because this logic is executed in the background. But I want to attach the requests's span to all jobs.

I think I can attach a span to each job in a loop, but I would like to place the start of the span inside the implementation of sending an HTTP request, but the problem is that only 1 context goes in there, not a context per job.

**Reply by dmathieu** [reply]
If what you're looking for is to have several spans stored within the context, that is not possible.

If you look at the link above, you will see we use [WithValue](https://pkg.go.dev/context#WithValue) to store information about the current span within the context, so any new span can know the parent span ID.
What your [mergectx.Value](https://github.com/SentimensRG/ctx/blob/master/mergectx/mergectx.go#L97-L102) does is read the value from the first context, and if it doesn't exist, read it from the second one.
So if you have multiple context with spans that you join, the first one found will be used as parent ID.
Having one span with multiple parents is not possible either.

It seems what you're interested in may be to be able to see the full duration of the work, from the beginning of your HTTP request to the job finishing.
In order to get that information, you shouldn't create spans earlier. Your spans should be created when the work they represent starts.
However, with parent/children spans, it's your tracing interface's job to compute the full duration of all the spans, including the child ones.
